### PR TITLE
Add RunInformers function

### PR DESCRIPTION
**Pull request by skaslev**
Add RunInfomers which is similar to the StartInformers function but allows for
users to get notified when the informers have finished running.

This function will be mainly used in tests to fix the race described in
knative/serving#5351

### CODE REVIEW

**Comment by skaslev** [comment]
See knative/serving#5795 for how this is intended to be used.

File: controller/controller.go [hunk 1]
```diff
@@ -445,6 +445,34 @@ func StartInformers(stopCh <-chan struct{}, informers ...Informer) error {
 	return nil
 }
 
+// RunInformers kicks off all of the passed informers and then waits for all of
+// them to synchronize. Returned channel will get closed after all informers
+// finish running.
+func RunInformers(stopCh <-chan struct{}, informers ...Informer) (<-chan struct{}, error) {
```
Comment (vagababov) [inline controller/controller.go]: `RunWithWait/RunAndWait`?

File: controller/controller.go [hunk 2]
```diff
@@ -445,6 +445,34 @@ func StartInformers(stopCh <-chan struct{}, informers ...Informer) error {
...
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(doneCh)
+	}()
+
+	for i, informer := range informers {
+		if ok := cache.WaitForCacheSync(stopCh, informer.HasSynced); !ok {
+			return doneCh, fmt.Errorf("failed to wait for cache at index %d to sync", i)
```
Comment (vagababov) [inline controller/controller.go]: do we want to close `doneCh` here?

File: controller/controller.go (hunk 1 above)
Comment (skaslev) [reply controller/controller.go]: I pushed a new implementation using `errgroup.Group` before having a chance to see those comments. Sorry about that.

I don't have strong opinion about the name as long as this does unflake the data race.

File: controller/controller.go [hunk 3]
```diff
@@ -445,6 +446,25 @@ func StartInformers(stopCh <-chan struct{}, informers ...Informer) error {
 	return nil
 }
 
+// RunInformers kicks off all of the passed informers in the given error group
+// and then waits for all of them to synchronize.
+func RunInformers(stopCh <-chan struct{}, grp *errgroup.Group, informers ...Informer) error {
```
Comment (vagababov) [inline controller/controller.go]: I liked previous approach better.
But the best would probably be to return a call back. Something like

```
func RI(stop, inf...) (cb, error) {
  var g errgroup.Group
  for i := range {
    i:=i
    grp.Go()
  }

  for ... {sync looooooop}
  return func() { wg.Wait()}, nil
```

Now the caller can do
```
ctx, cancel, inf := ....
cb, er := RunInformers(ctx.Done(), inf)
if err { cancel(); handleError }
defer func() {
  cancel()
  cb()
}
cancel()
```

**Comment by skaslev** [comment]
> Looks good, but we need a test that checks that the returned callback is invoked.

I don't quite understand. The callback should be called by users of `RunInformers` and the test is a user that either calls or doesn't the callback depending on does it want wait. 

Do you mean making sure that `waitInformers` is called when used in such a way?

```
			ctx, cancel, _ := rtesting.SetupFakeContextWithCancel(t)
			waitInformers := func() {}
			defer func() {
				cancel()
				waitInformers()
			}()

			waitInformers, err := controller.RunInformers(ctx.Done(), endpointsInformer.Informer())
			if err != nil {
				t.Fatalf("Failed to start informers: %v", err)
			}
```

**Comment by vagababov** [comment]
Yes, we need  to make sure it works (at least make sure we can call it and it won't hang).

**Comment by skaslev** [comment]
@vagababov, I added a new test. Please take another look.

File: controller/controller.go [hunk 4]
```diff
@@ -445,6 +445,27 @@ func StartInformers(stopCh <-chan struct{}, informers ...Informer) error {
...
+// RunInformers kicks off all of the passed informers and then waits for all of
+// them to synchronize. Returned function will wait for all informers to finish.
+func RunInformers(stopCh <-chan struct{}, informers ...Informer) (func(), error) {
+	var wg sync.WaitGroup
+	wg.Add(len(informers))
+	for _, informer := range informers {
+		informer := informer
+		go func() {
+			informer.Run(stopCh)
+			wg.Done()
```
Comment (vagababov) [inline controller/controller.go]:
```
defer wg.Done()
informer.Run(stopCh)
```

Comment (skaslev) [reply controller/controller.go]: Done.

**Pull request merged**